### DISCUSSION TITLE
Gator Oscillator - How do you plot on chart?

### DISCUSSION THREAD

**Discussion by jimtollan** [DaveSkender/Stock.Indicators#830] [Q&A] [answered]
Dave -so far, so good on using the chained functionality, does exactly as required. However, I have one small nagging issue in terms of using the Gator Oscillator in my charts. If I simply use the values Upper and Lower and assign it to Green and Red, then of course, it gives a completely false representation of the premise. 

In order to fully represent as required, I need to obviously have 4 values that can be colored i.e.:

- UpperExpanding
- UpperContracting
- LowerExpanding
- LowerContracting

To allow me to chain off of the `IEnumerable<GatorResult>`, i've created an extension method that creates these 4 members. However, I'm not entirely sure if I'm capturing the premise correctly as I end up with the occasional blank opposing bar in the bar chart. The code below shows what I'm doing here:

    using Skender.Stock.Indicators;
    using System;
    using System.Collections.Generic;

    namespace Wpf.Leverage
    {
        public class ExtendedGatorResult
        {
            public DateTime Date { get; set; }
            public double? UpperExpanding { get; set; }
            public double? LowerExpanding { get; set; }
            public double? UpperContracting { get; set; }
            public double? LowerContracting { get; set; }
        }
        public static class SkenderExtensions
        {
            public static IEnumerable<ExtendedGatorResult> ExtendGator(this IEnumerable<GatorResult> gatorResults)
            {
                List<ExtendedGatorResult> extendedGatorResults = new List<ExtendedGatorResult>();

                foreach (var item in gatorResults)
                {
                    var extendedGatorResult = new ExtendedGatorResult
                    {
                        Date = item.Date
                    };

                    if (item.UpperIsExpanding.HasValue)
                    {
                        extendedGatorResult.UpperExpanding = item.UpperIsExpanding.Value ? item.Upper : null;
                        extendedGatorResult.UpperContracting = !item.UpperIsExpanding.Value ? item.Upper : null;
                    }

                    if (item.LowerIsExpanding.HasValue)
                    {
                        extendedGatorResult.LowerExpanding = item.LowerIsExpanding.Value ? item.Lower : null;
                        extendedGatorResult.LowerContracting = !item.LowerIsExpanding.Value ? item.Lower : null;
                    }

                    extendedGatorResults.Add(extendedGatorResult);
                }

                return extendedGatorResults;
            }
        }
    }

Are you able to suggest a logic fix that would allow me to present the 4 members to my chart (devexpress) and have it display correctly?? 

The image below shows how it outputs it using the above extension method:

![gator-issue](https://user-images.githubusercontent.com/87867061/175170606-324c397c-5b42-424a-bf3b-3bd2bdf46cad.PNG)

Hope you can help...

cheers
jim

**Comment by DaveSkender** [comment, marked as answer]
@heavymanto just mentioned this same thing recently, in #819.  See this thread for conversation to help troubleshoot and clarify.

**Reply by jimtollan** [reply]
thanks dave!!

**Reply by DaveSkender** [reply]
I don’t have this one in our demo site yet, but do have it plotted in my private trading system.  It’s essentially the two sets of bars, one upper, one lower.  They’re green if `isExpanding`, or red if not.

```ts
// bar color
upperColor.push(m.upperIsExpanding ? 'darkGreen' : 'red');
lowerColor.push(m.lowerIsExpanding ? 'darkGreen' : 'red');
```

**Reply by jimtollan** [reply]
Cheers - -I think i can make that change to my extension method.. Will keep you posted.

thanks again
jim

**Reply by DaveSkender** [reply]
My charting system allows you to set a different color for each bar, so if you’re working with one that can only apply one color to the whole series, yes, you’d have to parse out 4 distinct series to get it to plot right.

**Reply by jimtollan** [reply]
to further add (to my woes maybe). Devexpress needs to have the 4 items individually defined in order to create 4 series (one for each representation).

So, I guess my issue is still present as my logic above is all about populating the 4 values with either null, Upper or Lower... If you have any insights, feel free to suggest 👯

**Reply by DaveSkender** [reply]
I don’t have any magic for that one.  Four series with `null` or `0`  to plug the holes is probably the hack I’d use here too, assuming timeseries axis aren’t an option for stacked bars in that tool.  You’d not need the plugs in some cases.

**Reply by jimtollan** [reply]
alas - i think the approach i've taken is probably the *cleanest* option. still feel i'm missing a trick, but it's not a deal breaker as this is all just for my own use anyway, so can circle back to it in due course. if/when i figure a better approach, i'll update this thread. thanks again for your inputs as ever!!

**Comment by DaveSkender** [comment]
If you like Gator, you might also like [Elder-Ray Index](https://daveskender.github.io/Stock.Indicators/indicators/ElderRay/#content).  I’ve been watching it lately in our [demo site](https://stock-charts.azurewebsites.net/).  It pretty good at showing overall short-term up/down trend direction.

![9EBAC861-E053-4996-B8FC-4A87BA39E687](https://user-images.githubusercontent.com/8432125/175179827-a69c8551-32e1-4857-9d64-53db0949465b.jpeg)

**Reply by jimtollan** [reply]
Yup, i already have that in my library :D.. that was fine as just two series values..

![elder-ray](https://user-images.githubusercontent.com/87867061/175179863-db55fe8a-487d-498f-9048-f11d11cd5402.PNG)

**Reply by jimtollan** [reply]
As you can tell, I'm a big fan of this library, but 2 I'd single out in particular are ADX and KJD. Those 2 just ooze perfect functionality, especially when used in combo with RSI/MACD and bollies etc..

**Reply by DaveSkender** [reply]
Yah, ADX and Stochastic Oscillator (aka KDJ) are some of my favorites as well.  SuperTrend, one of many ATR Trailing Stop variants, has worked quite well too (see [chart above](https://github.com/DaveSkender/Stock.Indicators/discussions/830#discussioncomment-3006052)) for general mid/longer-term trend monitoring.  Still going down!

**Reply by jimtollan** [reply]
yeah.. my supertrend/kjd (only possible thanks to a library called Skender.Indicators -don't know if you've heard of it  🤣 

![supertrand-kjd](https://user-images.githubusercontent.com/87867061/175182313-944dd6bb-db33-4636-aeae-7dd6e196b66c.PNG)
